<a href="https://colab.research.google.com/github/anasalek/Pyhton_hw/blob/main/Email_generator.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Создайте класс EmailGenerator, который использует LangChain и Groq API, чтобы создавать сообщения для персонализированной email-рассылки на основе входных данных пользователя. Класс должен иметь метод run(user_data: str) -> str, который принимает данные для генерации email'ов и возвращает текст сообщения для рассылки с необходимым форматирования. Используйте шаблоны промптов и цепочки langchain. Предусмотрите возможность генерации сообщений на основе автоматического выбора шаблонов email'ов для рекламы, новостной рассылки, корпоративных сообщений. Шаблоны генерируются и выбираются автоматически на основе запроса пользователя.

## Это я случайно все сделала без LangChain, c ним смотрите ниже

In [ ]:
# грок апи
from openai import OpenAI
import os
client = OpenAI(
    api_key="your_api-key",
    base_url="https://api.groq.com/openai/v1",
)

In [ ]:
response = client.responses.create(
    input="What color is cherry?",
    model="openai/gpt-oss-20b"
)
print(response.output_text)

**Cherry** is essentially a deep, vivid red—think of the bright, almost pink‑ish hue of a ripe, juicy cherry fruit.  
In web‑design terms the color is often represented by the hex code **#DE3163** (a warm, slightly purplish‑red).  

If you’re looking for variations, cherries can range from a lighter “pale cherry” (closer to a soft pink) to a darker “deep cherry” (near true crimson). The key is that unmistakable, eye‑catching red quality that makes the word “cherry” feel both sweet and bold.


In [ ]:
class EmailGenerator():
  '''
  создает персонализированную email-рассылку
  на основе данных пользователя:
  не клиент получает рекламу,
  клиент получает сводку событий,
  работник получает распоряжения
  '''
  def __init__(self):
    self = self


  def run(self, user_data: str) -> str:
      prompt = "Create a short corporate email. The content should depend on the user role: not_user gets adds about our new vacuum cleaner, user gets info about the company events, worker gets instructions what to do and where to be. Now create an email for: "
      response = client.responses.create(
      input= prompt + user_data,
      model="openai/gpt-oss-20b"
      )
      return response.output_text


In [ ]:
# вон как красиво с разметочкой генерит
EmailGenerator.run("user")

'**Subject:** Upcoming Company Events – Save the Dates!\n\nHi [Name],\n\nI hope you’re doing well. I’m pleased to share a quick rundown of the key events we’ve lined up for the next few weeks. These are great opportunities to stay in the loop, connect with our team, and explore new initiatives.\n\n| Event | Date & Time | Location | RSVP |\n|-------|-------------|----------|------|\n| **Virtual Product Launch – “EcoVac 360”** | Nov\u202f15,\u202f2025\u202f•\u202f10:00\u202fAM\u202fCET | Online (Zoom link) | [RSVP here](#) |\n| **Annual Stakeholder Meeting** | Nov\u202f22,\u202f2025\u202f•\u202f9:00\u202fAM | HQ – Conference Room A | [RSVP here](#) |\n| **Web Development Workshop** | Nov\u202f30,\u202f2025\u202f•\u202f2:00\u202fPM | Online (Teams) | [RSVP here](#) |\n\nPlease let me know if you can attend any of these sessions or if you’d like additional information. We’ll circulate detailed agendas and supporting materials a week before each event.\n\nThank you for being a valued part o

**Subject:** Upcoming Company Events – Save the Dates!

Hi [Name],

I hope you’re doing well. I’m pleased to share a quick rundown of the key events we’ve lined up for the next few weeks. These are great opportunities to stay in the loop, connect with our team, and explore new initiatives.

| Event | Date & Time | Location | RSVP |
|-------|-------------|----------|------|
| **Virtual Product Launch – “EcoVac 360”** | Nov 15, 2025 • 10:00 AM CET | Online (Zoom link) | [RSVP here](#) |
| **Annual Stakeholder Meeting** | Nov 22, 2025 • 9:00 AM | HQ – Conference Room A | [RSVP here](#) |
| **Web Development Workshop** | Nov 30, 2025 • 2:00 PM | Online (Teams) | [RSVP here](#) |

Please let me know if you can attend any of these sessions or if you’d like additional information. We’ll circulate detailed agendas and supporting materials a week before each event.

Thank you for being a valued part of our community. Looking forward to seeing you there!

Best regards,

**Alex Martinez**  
Director of Corporate Communications  
[Company Name]  
(555) 123‑4567 | alex.martinez@company.com

---

In [ ]:
EmailGenerator.run("not_user")

'**Subject:** Experience the Future of Cleaning with Our New Vacuum Cleaner  \n\nDear Customer,\n\nWe’re excited to introduce our latest innovation – the **AirClean 360 Ultra** vacuum cleaner. Designed with cutting‑edge technology, it delivers:\n\n- **Powerful suction** that tackles dust, pet hair, and fine particles with ease.  \n- **Ultra‑quiet operation** for a peaceful home environment.  \n- **Smart home integration** that lets you schedule cleaning from your phone or smart assistant.  \n\nFor a limited time, enjoy a **15\u202f% launch discount** when you order before [date]. Click the link below to learn more and take advantage of this exclusive offer.\n\n[Shop Now – AirClean 360 Ultra]\n\nThank you for choosing us to help keep your home cleaner and healthier.\n\nBest regards,\n\nThe AirClean Team  \n[Company Name] | www.companywebsite.com | 1‑800‑555‑1234'

**Subject:** Experience the Future of Cleaning with Our New Vacuum Cleaner  

Dear Customer,

We’re excited to introduce our latest innovation – the **AirClean 360 Ultra** vacuum cleaner. Designed with cutting‑edge technology, it delivers:

- **Powerful suction** that tackles dust, pet hair, and fine particles with ease.  
- **Ultra‑quiet operation** for a peaceful home environment.  
- **Smart home integration** that lets you schedule cleaning from your phone or smart assistant.  

For a limited time, enjoy a **15 % launch discount** when you order before [date]. Click the link below to learn more and take advantage of this exclusive offer.

[Shop Now – AirClean 360 Ultra]

Thank you for choosing us to help keep your home cleaner and healthier.

Best regards,

The AirClean Team  
[Company Name] | www.companywebsite.com | 1‑800‑555‑1234

In [ ]:
EmailGenerator.run("worker")

'**Subject:**\u202fWeekly Assignment & Location – [Date]\n\nHi\u202f[Worker Name],\n\nThis week, you’re scheduled for the following tasks and locations:\n\n| Date | Time | Task | Location |\n|------|------|------|----------|\n| Mon,\u202fOct\u202f21 | 08:00‑12:00 | Inventory audit – Zone A | Warehouse – Section 3 |\n| Mon,\u202fOct\u202f21 | 13:00‑17:00 | Restocking – Zone B | Warehouse – Section 5 |\n| Tue,\u202fOct\u202f22 | 09:00‑12:30 | Equipment inspection – Section 1 | Workshop – Room 2 |\n| Tue,\u202fOct\u202f22 | 14:00‑17:00 | Packing – Zone C | Packing Hall – Floor 1 |\n\n**Key points:**\n- Bring your badge and the weekly checklist.\n- Report to the supervisor at the specified time.\n- Use the provided safety gear for all inspections.\n- Submit the completed checklist to the supervisor by 18:00 each day.\n\nIf you have any questions or need a change in schedule, please let me know by end of day Monday.\n\nThanks for your hard work.\n\nBest regards,\n\n[Your Name]  \nOperations

**Subject:** Weekly Assignment & Location – [Date]

Hi [Worker Name],

This week, you’re scheduled for the following tasks and locations:

| Date | Time | Task | Location |
|------|------|------|----------|
| Mon, Oct 21 | 08:00‑12:00 | Inventory audit – Zone A | Warehouse – Section 3 |
| Mon, Oct 21 | 13:00‑17:00 | Restocking – Zone B | Warehouse – Section 5 |
| Tue, Oct 22 | 09:00‑12:30 | Equipment inspection – Section 1 | Workshop – Room 2 |
| Tue, Oct 22 | 14:00‑17:00 | Packing – Zone C | Packing Hall – Floor 1 |

**Key points:**
- Bring your badge and the weekly checklist.
- Report to the supervisor at the specified time.
- Use the provided safety gear for all inspections.
- Submit the completed checklist to the supervisor by 18:00 each day.

If you have any questions or need a change in schedule, please let me know by end of day Monday.

Thanks for your hard work.

Best regards,

[Your Name]  
Operations Supervisor  
[Company Name]  
[Phone] | [Email]

## То же самое с ChatGroq и PromptTemplate, потому что кто-то не сразу вник, что это langchain chat_model, а не цепочка...

In [ ]:
pip install langchain_community

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 36.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 32.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 467.2/467.2 kB 14.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.7/64.7 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 2.6 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.32.4
    Uninstalling requests-2.32.4:
      Successfully uninstalled requests-2.32.4
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.3.79
    Uninstalling langchain-core-0.3.79:
      Successfully uninstalled langchain-core-0.3.79
  Attempting uninstall: langchain-text-splitters
    Found existing installation: langchain-text-splitters 0.3.11
    Uninstalling langchain-text-splitters-0.3.11:
      Successfully uninstalled langchain-text-splitters-0.3.11
ERROR: pip's dependency resolver

In [ ]:
pip install -U langchain-groq

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.4/135.4 kB 5.7 MB/s eta 0:00:00


In [ ]:
from langchain_groq import ChatGroq
from langchain_community.llms import OpenAI
from langchain_core.prompts import PromptTemplate

class EmailGenerator():
  '''
  создает персонализированную email-рассылку
  на основе данных пользователя:
  не клиент получает рекламу,
  клиент получает сводку событий,
  работник получает распоряжения
  '''
  def __init__(self):
    self = self


  def run(self, user_data: str) -> str:
    prompt_template = "Create a short corporate email. The content should depend on the user role: not_client gets adds about our new vacuum cleaner, client gets info about the company events, worker gets instructions what to do and where to be. Now create an email for: {user}"
    prompt = PromptTemplate(
        input_variables=["user_data"], template=prompt_template
    )

    llm = ChatGroq(
        api_key="your_api-key",
        model="llama-3.1-8b-instant", # ну вот модель еще другая
        temperature=0.0,
        max_retries=2,
    )
    chain = prompt | llm
    return chain.invoke(user_data).content


In [ ]:
EmailGenerator.run("client")

"Subject: Upcoming Company Events\n\nDear [Client Name],\n\nWe hope this email finds you well. As we continue to grow and expand our services, we would like to invite you to join us for some exciting company events. These events are a great opportunity to network with our team, learn more about our products and services, and get a behind-the-scenes look at our company culture.\n\nHere are the details of the upcoming events:\n\n- **Company BBQ**: Join us on Friday, October 27th at 12:00 PM at our office park for a fun-filled afternoon of food, games, and networking. We will have a variety of delicious food options, a photo booth, and a raffle with exciting prizes.\n- **Product Showcase**: On November 10th at 2:00 PM, we will be hosting a product showcase at our headquarters. This event will feature our latest innovations, including our new vacuum cleaner, and will provide a great opportunity to learn more about our products and services.\n- **Holiday Party**: Mark your calendars for Dec

In [ ]:
EmailGenerator.run("worker")

"Subject: Important Update for Today's Tasks\n\nDear [Worker's Name],\n\nAs we approach the end of the quarter, it's essential that we stay focused on our goals and meet our deadlines. Below are the key tasks and locations you need to be aware of for today:\n\n- **Meeting with Marketing Team**: You are scheduled to meet with the marketing team at 10:00 AM in Conference Room A to discuss the upcoming product launch. Please come prepared with any questions or concerns you may have.\n- **Project Update**: You will be working on the Smith account today, focusing on the Johnson report. Please ensure that you complete the report by 2:00 PM and submit it to the project manager for review.\n- **Training Session**: At 3:00 PM, you will be attending a training session on our new CRM system in Training Room B. This session is mandatory, and it's essential that you attend to ensure you're up-to-date on the latest software updates.\n\nIf you have any questions or concerns, please don't hesitate to 

Subject: Upcoming Company Events

Dear [User's Name],

We hope this email finds you well. As a valued member of our community, we would like to invite you to our upcoming company events. These events are a great opportunity to meet our team, learn more about our company culture, and get a glimpse into our latest projects.

**Upcoming Events:**

1. **Team Lunch**: Join us on Friday, October 20th at 12:00 PM in the conference room for a team lunch. We will be serving a variety of cuisines and there will be plenty of opportunities to mingle with your colleagues.
2. **Company Update**: On Wednesday, October 25th at 2:00 PM, our CEO will be hosting a company update session in the auditorium. This is a great chance to learn about our company's progress and vision for the future.
3. **Networking Event**: On Thursday, November 2nd at 6:00 PM, we will be hosting a networking event at a local restaurant. This is a great opportunity to meet new people and make connections in the industry.

**Registration:**

To register for any of these events, please click on the following link: [Registration Link]. If you have any dietary restrictions or special requests, please let us know in the comments section.

We look forward to seeing you at these events!

Best regards,

[Your Name]
[Company Name]

Шаблоны промптов

In [ ]:
# вот еще так можно
from langchain_core.prompts import ChatPromptTemplate

prompt_template = ChatPromptTemplate([
    ("system", "Create a short corporate email"),
    ("user", "Create a short corporate email for {user}")
])
# и тогда будет так (но зачем)
prompt_template.invoke({"user": "worker"})

## SequentialChain


Я немного не поняла, зачем конкретно тут LangChain с цепочками, если получилось и без него... Поэтому добавила цепочку для определения роли пользователя на случай, если мы не сами задаем user_data, а если модели это надо понять из сообщений пользователя самостоятельно. Поэтому SequentialChain

In [4]:
pip install langchain_community

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 24.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.7/64.7 kB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 4.9 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.32.4
    Uninstalling requests-2.32.4:
      Successfully uninstalled requests-2.32.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests==2.32.4, but you have requests 2.32.5 which is incompatible.
langchain 0.3.27 requires langchain-core<1.0.0,>=0.3.72, but you have langchain-core 1.0.0 which is incompatible.
langchain 0.3.27 requires langchain-text-splitters<1.0.0,>=0.3.9, but you have langchain-text-splitters 1.0.0 which is incompatible.


In [2]:
pip install langchain_openai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.9/81.9 kB 2.7 MB/s eta 0:00:00


In [1]:
pip install langchain_classic

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 11.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 467.2/467.2 kB 20.4 MB/s eta 0:00:00
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.3.79
    Uninstalling langchain-core-0.3.79:
      Successfully uninstalled langchain-core-0.3.79
  Attempting uninstall: langchain-text-splitters
    Found existing installation: langchain-text-splitters 0.3.11
    Uninstalling langchain-text-splitters-0.3.11:
      Successfully uninstalled langchain-text-splitters-0.3.11
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langchain 0.3.27 requires langchain-core<1.0.0,>=0.3.72, but you have langchain-core 1.0.0 which is incompatible.
langchain 0.3.27 requires langchain-text-splitters<1.0.0,>=0.3.9, but you have langchain-text-splitters 1.0.0 which is incompatible.


In [52]:
from langchain_classic.chains import SequentialChain, LLMChain
from langchain_core.prompts import PromptTemplate
from langchain_openai import ChatOpenAI

class EmailGenerator():
    '''
    создает персонализированную email-рассылку
    на основе данных пользователя:
    не клиент получает рекламу,
    клиент получает сводку событий,
    работник получает распоряжения
    '''
    def __init__(self):

        self.llm = ChatOpenAI(
        api_key="your_api-key",
        base_url="https://api.groq.com/openai/v1",
        model="openai/gpt-oss-20b",
        temperature=0.0,
        max_retries=2,
        )

        self.role_check =  LLMChain(
            llm=self.llm,
            prompt=PromptTemplate(
                template="Analyse the user message {user_data} and guess their role: not_client, client, worker. Write nothing but the role",
                input_variables= ["user_data"],
            ),
            output_key="user_role"
        )

        self.write_email = LLMChain(
            llm=self.llm,
            prompt=PromptTemplate(
                input_variables=['user_role'],
                template="Create a short corporate email. The content should depend on the user role: not_client gets adds about our new vacuum cleaner, client gets info about the company events, worker gets instructions what to do and where to be. Now create an email for: {user_role}"
            ),
            output_key="email"
        )


        self.sequential_chain = SequentialChain (
            chains=[self.role_check, self.write_email],
            input_variables=["user_data"],
            output_variables=["user_role","email"],
        )

    def run(self, user_data: str) -> str:
        result = self.sequential_chain.invoke({user_data})
        return result["email"]

In [54]:
emreg = EmailGenerator()
letter = emreg.run(user_data="Where do I find the manager?")
print(letter)

**Subject:** Weekly Operations Brief – Monday, October 28

Hi [Name],

Below are your key assignments for this week:

| Day | Task | Location | Notes |
|-----|------|----------|-------|
| **Mon** | Complete inventory audit of the 3rd‑floor warehouse | 3rd‑floor – Dock A | Bring the barcode scanner; report discrepancies to logistics by 10 AM. |
| **Tue** | Assist with the new HVAC installation | Main lobby – Conference Room B | Follow the safety protocol; wear the provided hard hat and gloves. |
| **Wed** | Restock office supplies in the supply closet | 2nd‑floor – Supply Closet | Check the back‑order list; update the spreadsheet in the shared drive. |
| **Thu** | Conduct routine maintenance on the production line A | Production Zone 1 | Use the maintenance log; notify the supervisor if any component needs replacement. |
| **Fri** | Clean and sanitize all break rooms | 1st‑floor – Break Rooms 1 & 2 | Ensure disinfectant wipes are replenished; complete the checklist in the app. |

**Impo